datafusion/core/src/physical_optimizer/enforcement.rs

```diff
@@ -0,0 +1,2001 @@
+                "ProjectionExec: expr=[a@0 as a1, b@1 as b1, c@2 as c1]",
+                "ParquetExec: limit=None, partitions=[x], projection=[a, b, c, d, e]",
+                "HashJoinExec: mode=Partitioned, join_type=Inner, on=[(Column { name: \"c\", index: 2 }, Column { name: \"c1\", index: 2 }), (Column { name: \"b\", index: 1 }, Column { name: \"b1\", index: 1 }), (Column { name: \"a\", index: 0 }, Column { name: \"a1\", index: 0 })]",
+                "RepartitionExec: partitioning=Hash([Column { name: \"c\", index: 2 }, Column { name: \"b\", index: 1 }, Column { name: \"a\", index: 0 }], 10)",
+                "ParquetExec: limit=None, partitions=[x], projection=[a, b, c, d, e]",
+                "RepartitionExec: partitioning=Hash([Column { name: \"c1\", index: 2 }, Column { name: \"b1\", index: 1 }, Column { name: \"a1\", index: 0 }], 10)",
+                "ProjectionExec: expr=[a@0 as a1, b@1 as b1, c@2 as c1]",
+                "ParquetExec: limit=None, partitions=[x], projection=[a, b, c, d, e]",
+            ];
+
+            assert_plan_txt!(expected, reordered);
+        }
+
+        Ok(())
+    }
+
+    #[test]
+    fn multi_smj_joins() -> Result<()> {
+        let left = parquet_exec();
+        let alias_pairs: Vec<(String, String)> = vec![
+            ("a".to_string(), "a1".to_string()),
+            ("b".to_string(), "b1".to_string()),
+            ("c".to_string(), "c1".to_string()),
+            ("d".to_string(), "d1".to_string()),
+            ("e".to_string(), "e1".to_string()),
+        ];
+        let right = projection_exec_with_alias(parquet_exec(), alias_pairs);
+
+        // SortMergeJoin does not support RightSemi and RightAnti join now
+        let join_types = vec![
+            JoinType::Inner,
+            JoinType::Left,
+            JoinType::Right,
+            JoinType::Full,
+            JoinType::LeftSemi,
+            JoinType::LeftAnti,
+        ];
+
+        // Join on (a == b1)
+        let join_on = vec![(
+            Column::new_with_schema("a", &schema()).unwrap(),
+            Column::new_with_schema("b1", &right.schema()).unwrap(),
+        )];
+
+        for join_type in join_types {
+            let join =
+                sort_merge_join_exec(left.clone(), right.clone(), &join_on, &join_type);
+            let join_plan =
+                format!("SortMergeJoin: join_type={}, on=[(Column {{ name: \"a\", index: 0 }}, Column {{ name: \"b1\", index: 1 }})]", join_type);
+
+            // Top join on (a == c)
+            let top_join_on = vec![(
+                Column::new_with_schema("a", &join.schema()).unwrap(),
+                Column::new_with_schema("c", &schema()).unwrap(),
+            )];
+            let top_join = sort_merge_join_exec(
+                join.clone(),
+                parquet_exec(),
+                &top_join_on,
+                &join_type,
+            );
+            let top_join_plan =
```

Comment:
Something I don't see in these plans (or in the code) is the fact that if you take several sorted streams and run them through `RepartitionExec` I don't think the output remains sorted. I think these plans are ok, as they sort after the repartition, but in general that is something to watch out for